zkevm-circuits/src/evm_circuit/execution/mul.rs

```diff
@@ -29,30 +40,65 @@ impl<F: Field> ExecutionGadget<F> for MulGadget<F> {
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
 
-        let a = cb.query_word();
-        let b = cb.query_word();
+        let is_mul = (OpcodeId::DIV.expr() - opcode.expr())
+            * (OpcodeId::MOD.expr() - opcode.expr())
+            * F::from(8).invert().unwrap();
+        let is_div = (opcode.expr() - OpcodeId::MUL.expr())
+            * (OpcodeId::MOD.expr() - opcode.expr())
+            * F::from(4).invert().unwrap();
+        let is_mod = (opcode.expr() - OpcodeId::MUL.expr())
+            * (opcode.expr() - OpcodeId::DIV.expr())
+            * F::from(8).invert().unwrap();
+        let mul_add_words = MulAddWordsGadget::construct(cb);
+        let divisor_is_zero = IsZeroGadget::construct(cb, sum::expr(&mul_add_words.b.cells));
+        let lt_word = LtWordGadget::construct(cb, &mul_add_words.c, &mul_add_words.b);
 
-        // Pop a and b from the stack, push product on the stack
-        cb.stack_pop(a.expr());
-        cb.stack_pop(b.expr());
-        let mul_words = MulWordsGadget::construct(cb, a, b);
-        cb.stack_push(mul_words.product().expr());
+        // Pop a and b from the stack, push result on the stack
+        // The first pop is multiplier for MUL and dividend for DIV/MOD
+        // The seoncd pop is multiplicand for MUL and divosr for DIV/MOD
+        // The push is product for MUL, quotient for DIV, and residue for MOD
+        // Note that for DIV/MOD, when divisor == 0, the push value is also 0.
+        cb.stack_pop(select::expr(
+            is_mul.clone(),
+            mul_add_words.a.expr(),
+            mul_add_words.d.expr(),
+        ));
+        cb.stack_pop(mul_add_words.b.expr());
+        cb.stack_push(
+            is_mul.clone() * mul_add_words.d.expr()
+                + is_div * mul_add_words.a.expr() * (1.expr() - divisor_is_zero.expr())
+                + is_mod * mul_add_words.c.expr() * (1.expr() - divisor_is_zero.expr()),
+        );
+
+        // Constraint for MUL case
+        cb.add_constraint(
+            "c == 0 for opcode MUL",
+            is_mul.clone() * sum::expr(&mul_add_words.c.cells),
+        );
+
+        // Constraints for DIV and MOD cases
+        cb.condition(1.expr() - is_mul, |cb| {
+            cb.add_constraint(
+                "residue < divisor when divisor != 0 for opcode DIV/MOD",
+                (1.expr() - lt_word.expr()) * (1.expr() - divisor_is_zero.expr()),
+            );
+            cb.add_constraint("overflow == 0 for opcode DIV/MOD", mul_add_words.overflow());
```

Comment:
there is a `require_zero` method you can use here that is more clear imo. all it does is call `add_constraint`.

Reply:
Updated